add single image uploads when editing submissions

dodumosu opened the pull request:
adds support for uploading images for observer submissions (checklists, incidents, surveys).

dodumosu committed:
feat: add image upload fields to form and template

dodumosu committed:
chore: change button colour and leave text

dodumosu committed:
feat: add single image upload

this commit adds upload of images while editing a submission.
currently, it is limited to editing critical incidents,
observer checklists or surveys.
this means adding a new critical incident doesn't upload any images

dodumosu committed:
feat: support image upload for incidents

takinbo commented on apollo/frontend/templates/frontend/macros/submission_field_macros.html:
```diff
@@ -8,6 +8,11 @@
+{% elif field_type == 'image' %}
+<div>
+{{ form_field(class_=klass_, disabled=disabled, hidden=true, accepts="image/jpeg, image/png") }}
+<input type="button" class="btn btn-secondary" data-context="select_button" value="{% trans %}Select file{% endtrans %}">
```
button labels should be in [Title Case](https://dictionary.cambridge.org/dictionary/english/title-case) to be consistent with the rest of the application.

takinbo replied:
Here's a better alternative:
```html
<label class="btn btn-secondary w-100">Choose File<input type="file" accept=".jpg,.png,image/jpeg,image/png" hidden /></label>
```

takinbo commented on apollo/submissions/forms.py:
```diff
@@ -146,6 +147,17 @@ def make_submission_edit_form_class(event, form):
+                        validators=[
+                            FileAllowed(
+                                ['jpg', 'png'],
+                                _('Only JPG/PNG images supported')
```
Change to: "Please select a PNG or JPEG image file to upload"

takinbo commented on apollo/submissions/views_submissions.py:
```diff
@@ -197,7 +198,7 @@ def submission_list(form_id):
             # only change the submission type if we are aggregating data
             # and are not tracking conflicts
-            if mode == 'aggregated' and form.untrack_data_conflicts == True:
+            if mode == 'aggregated' and form.untrack_data_conflicts == True:    # noqa
```
Two spaces should be between `:` and `#`

takinbo commented on apollo/submissions/views_submissions.py:
```diff
@@ -702,13 +732,19 @@ def submission_edit(submission_id):
         else:
             master_form = None
 
+        if master_form is None:
+            num_forms = len(sibling_forms) + 1
+        else:
+            num_forms = len(sibling_forms) + 2
+
```
Unnecessary if it's only going to be used for one type of field. I suggest two options:

1. You remove this and compute the colspan in the template or
2. You change all computations for colspan in the template

takinbo replied:
Unnecessary if it's only going to be used for one type of field. I suggest two options:

1. You remove this and compute the colspan in the template or
2. You change all computations for colspan in the template

takinbo reviewed (changes requested):
Rather than replacing whatever file was uploaded, this continues to add files to the question.

takinbo reviewed (changes requested):
Rather than replacing whatever file was uploaded, this continues to add files to the question.

dodumosu committed:
chore: change button text

dodumosu committed:
chore: change error text

dodumosu committed:
chore: change spacing

dodumosu committed:
chore: change casing

dodumosu committed:
fix: preserve image UUIDs on submit

this commit preserves the UUIDs of uploaded
images. checking the data property of a file field
in a `FlaskForm` instance seems to return `None` even if a file was sent

dodumosu committed:
chore: move colspan computation to template

takinbo commented on apollo/frontend/templates/frontend/macros/submission_field_macros.html:
```diff
@@ -8,6 +8,11 @@
 {% elif field_type == 'select' %}
 {{ form_field(class_="custom-select " + klass_, disabled=disabled) }}
+{% elif field_type == 'image' %}
+<div>
+{{ form_field(class_=klass_, disabled=disabled, hidden=true, accepts="image/jpeg, image/png") }}
+<input type="button" class="btn btn-secondary" data-context="select_button" value="{% trans %}Choose File{% endtrans %}">
+</div>
```
```html
<label class="btn btn-secondary w-100">Choose File<input type="file" accept=".jpg,.png,image/jpeg,image/png" hidden /></label>
```

dodumosu committed:
chore: change upload field implementation

dodumosu committed:
chore: add keyboard support to file upload

takinbo commented:
Squash all the commits into one in preparation to get it merged.
